xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -194,9 +206,6 @@ protected void handleLdsResponse(String versionInfo, List<Any> resources, String
 
       // LdsUpdate parsed successfully.
       parsedResources.put(listenerName, new ParsedResource(ldsUpdate, resource));
-      if (ldsUpdate.rdsName != null) {
-        retainedRdsResources.add(ldsUpdate.rdsName);
```

Comment:
My comment is really for lines 220-225 since I traced how `retainedRdsResources` is used. This is the first time I am looking at RDS processing code so I might be missing something. 

- why would you call `onAbsent()` on certain RDS subscribers just because you didn't find an RDS resource in the current LDS response? For server side (at least) we just get the Listener that was requested so the control plane may/will not send all Listeners in a single LDS response and you cannot conclude an RDS is absent just because you didn't find it in this one

- would `containedRdsResources` be a better name? I am trying to understand in what way are these RDS resources "retained". Same comment for other names with `retained...`

Reply:
> For server side (at least) we just get the Listener that was requested so the control plane may/will not send all Listeners in a single LDS response and you cannot conclude an RDS is absent just because you didn't find it in this one

LDS is state-of-the-world protocol, the control plane _should_ send all Listener(s) that the client is requesting. We don't care anything that is not related to what we are requesting for. So from this single XdsClient's perspective, there is no difference between having only the requested Listener(s) or there are thousands of other Listeners. Even if the control plane send us all of them, unrequested ones are just ignored.

`retainedRdsResources` is for deleting previously received RDS resources. As you know RDS is an incremental protocol, say if the control plane wants to send {A, B}, it may send them as {A} and {B} in separate RDS responses. But how can the control plane delete resource A? It is not able to do the deletion with RDS protocol itself (e.g., resending {B} doesn't tell the client that A is deleted). It needs to work with the LDS protocol (i.e., its parent protocol). That's why every RDS resource must be referenced by some LDS resource. In this example, both A and B must have been referenced by some Listener (via the [RDS field in HttpConnectionManager](https://github.com/grpc/grpc-java/blob/814655cdde5797854289ce5c2ec3e7b80ce0cf44/xds/third_party/envoy/src/main/proto/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L290)). With that, the client needs to keep track of RDS resources referenced by received Listeners. If the Listener we received discard the reference of RDS resource A, this is effectively the same as control plane telling us to delete RouteConfiguration A. Again, everything should be within the scope of Listener(s) that we are requesting for. For example, if some RouteConfiguration X is referenced by Listener Y, but we are not interested in Listener Y. We will never need/use RouteConfiguration X. So even if the control plane doesn't give us Listener Y, we don't need to consider the existence of RouteConfiguration X.

> would containedRdsResources be a better name?

That's just a local variable, doesn't bother much. Also, it doesn't seem to be a bad name, as it does refer to the set of RDS resources that is being referenced by the received Listener(s) (i.e., they are not deleted).

Reply:
> ...For example, if some RouteConfiguration X is referenced by Listener Y, but we are not interested in Listener Y. We will never need/use RouteConfiguration X....

Okay I thought a Route resource could be referenced by more than one Listener. So never mind...

Reply:
> > ...For example, if some RouteConfiguration X is referenced by Listener Y, but we are not interested in Listener Y. We will never need/use RouteConfiguration X....
> 
> Okay I thought a Route resource could be referenced by more than one Listener. So never mind...

Hmmm, based on this https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#deleting-resources it seems I was right. It says 

"For example, when the client receives an LDS update removing a Listener that was previously pointing to RouteConfiguration A, if no other Listener is pointing to RouteConfiguration A, then the client may delete A."

Now that being the case, the code doesn't match the above spec. Do you agree or should I illustrate with an example?

Reply:
The code is implementing the spec, what do you think it goes wrong? We collect all RDS resources in the Listener(s), and then only retain those RDS resources.

Reply:
Note that Listener processing in gRPC is very different from Envoy Listener processing. ApiListener (client side) and the TcpListener for the queried resource (server side) are gRPC specific things whereas in Envoy it gets all the top level resources (Listener and Cluster - but let's just focus on Listeners for this discussion) in a single response. So in Envoy's version of `handleLdsResponse` it will receive and process all Listeners so any unreferenced RDS resources are truly deleted after that processing.

But in gRPC the client side queries for a particular ApiListener (e.g. `xds:///example-service`) and control plane returns only that ApiListener and HCM and so on. Let's say this one references RDS X. A different client in the same app could be asking for `xds:///foobar-service` and that one would reference (indirectly) also RDS X. So X is a shared RDS resource between the 2 ApiListener. If the next Listener update for either one stops using RDS X or the if the Listener itself is deleted that doesn't mean RDS X should be considered deleted in your code because the other Listener is still referencing it.

The exact same thing is also happening on the server side. The control plane returns only the requested Listener and related RDS resources - so in the `handleLdsResponse` based on processing that single TcpListener you cannot consider some shared RDS resources as deleted.

Reply:
Wait, there is only one XdsClient by all servers/channels in the same application process. That is, all xDS resource management is handled by the single XdsClient. So in your case, if your application is running two gRPC servers/channels, one requesting for `xds:///example-service` Listener and the other requesting for `xds:///foobar-service` Listener. The single XdsClient sends the LDS request with {`xds:///example-service`, `xds:///foobar-service`} as the resource names. If the received Listeners both reference RDS X, two RDS resource watchers (one by each server/channel) will be added to the XdsClient requesting for the same RDS resource. Then when receiving the RouteConfiguration X, both watchers will be notified. After that, if receiving a new LDS response with Listener `xds:///example-service` changing to reference RDS Y instead, RouteConfiguration X is not deleted (i.e., XdsClient will not notify either RDS watcher that the resource it is subscribing to is deleted). Then the server/channel subscribing to `Listener `xds:///example-service` will cancel its watcher for RDS X and add a watcher for RDS Y. Now the XdsClient will request RDS resources {X, Y}.

Reply:
Okay to extend your example for the code
```
    for (String resource : rdsResourceSubscribers.keySet()) {
      if (!retainedRdsResources.contains(resource)) {
        ResourceSubscriber subscriber = rdsResourceSubscribers.get(resource);
        subscriber.onAbsent();
      }
    }
```
When a new LDS response for `xds:///example-service` replaces RDS X with RDS Y, the `retainedRdsResources` set will contain Y and *not* X. So when the loop finds the subscriber for X and the condition `!retainedRdsResources.contains(resource)` is true for resource X it will call `subscriber.onAbsent()` for the subscriber of X which will also inform the watcher related to `xds:///foobar-service` that the resource X is absent i.e. deleted. Did I get it wrong?

Reply:
That's right. But the watcher for `xds:///foobar-service` is still there. If the Listener for `xds:///foobar-service` still references RDS X, `retainedRdsResources` will contain RDS X.

Reply:
> That's right. But the watcher for `xds:///foobar-service` is still there. If the Listener for `xds:///foobar-service` still references RDS X, `retainedRdsResources` will contain RDS X.

Oh, so this is based on the assumption that the control plane will include all Listener resources in the LDS response even if only the listener for `xds:///example-service` changed. I looked at https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#client-behavior to validate the assumption and it really discusses the Envoy variant (where Envoy always has a wildcard subscription to Listeners) so it is not clear that for the gRPC variant - where gRPC always requests a named Listener resource (both for the client and server side) that the control plane is obligated to include all subscribed Listener resources in all LDS responses. But let me know if I missed anything.

Reply:
Why does unrelated resource matter? Say there are 1000 Listeners in total, but the client is only requesting for a specific one, say Listener A. There is no difference, from the client's perspective, whether the control plane sends you Listener A only or the entire 1000 Listeners. As the client will just ignore the rest 999. If Listener A references RDS resource X, and the rest 999 Listeners all reference RDS resource Y. Why does the client need to know the existence of RDS Y? It will never use it. So from the client's perspective, there is no RDS Y.

Reply:
Okay. Discussed offline and this is not an issue.